### DISCUSSION TITLE
FastAPI events behaving weird

### DISCUSSION THREAD

**Discussion by mabdullahadeel** [fastapi/fastapi#5797] [Questions]
Hi there,

I don't where I am doing things wrong but `shutdown` event in my application is not behaving as expected.
So, I am using `shoutdown` event to unsubscribe from the SNS topic (a `sync` call using `boto3`). 

Here is the actual snippet

When the application is shutdown `Ctrl+C`

```python
@app.on_event("shutdown")
def app_shutdown():
    sns_notification_subscriber.unsubscribe()
```

here is snippet for `sns_notification_subscriber.unsubscribe()`

```python
def unsubscribe(self) -> None:
        print("unsubscribe")     # <- first print statement
        res = self.sns_client.unsubscribe(
            SubscriptionArn=self.subscription_arn
        )
        print("suceessfully unsubscribed from topic")     # <- second print statement
```

The behavior is that the first print is executed but the second print never gets executed and the application stops.

### Interesting thing I noticed while debugging

my startup event

```python
@app.on_event("startup")
async def app_init():
    await connection_manager.connect()    # <- database connection
    sns_notification_subscriber.subscribe(    # <- SNS subscription
        protocol="http" if settings.DEBUG else "https",
        endpoint='endpoint/to/the/server'
    )
```

I was using this 👇 script to start the whole application in dev (for convenience)

```python
# main.py

import uvicorn
import sys

if __name__ == "__main__":
    is_prod = '--prod' in sys.argv
        
    uvicorn.run(
        "app.main:app",
        host="127.0.0.1",
        port=8080,
        reload=not is_prod,
    )
```

when I use this 👆 script to start the application (by running `python main.py` from the terminal) `startup` event works just fine but `shutdown` event does not work properly.

### But
when I run
```sh
uvicorn app:app --reload
```
to start the server, the `startup` event does not work (it does not connect to `db` and SNS) but the `shutdown` event works.

Why is this?
Am I doing something silly?

Thanks!

**Comment by iudeen** [comment]
Are you on Windows? 

Also, it's not recommended (and a bad practice) to use reload on production scenarios. JFYI

**Reply by mabdullahadeel** [reply]
@iudeen yeah, I am on windows.

Not sure about what you meant by
> reload in production scenario

**Reply by mabdullahadeel** [reply]
I understand that I shouldn't pass --reload flag or equivalent in prod.
But not sure how it's related here

**Comment by jgould22** [comment]
Some things to think about

- Boto3 is not async so it will block the event loop and could cause problems 
- Are you confirming the subscription / unsubscribe and or connection to the db at the other end
- Can you be more descriptive when you say `second print never gets executed and the application stops.`, what does it mean for the application to "stop"  ?
- Are you debugging based on print statements and are you sure they have been flushed?
- `The shutdown handlers will run once all connections have been closed, and any in-process background tasks have completed.` are you holding open a connection or have long running background tasks?

**Reply by mabdullahadeel** [reply]
@jgould22 

<hr>

> Boto3 is not async so it will block the event loop and could cause problems

I am aware that boto3 is synchronous. To check if that's causing the issue, I have tried wrapping boto3 unsubscribe call in `sync_to_async` from `asgiref`. But that too did not work.
<hr>

> Are you confirming the subscription / unsubscribe and or connection to the db at the other end

Yes, database connections and SNS subscriptions are being confirmed since I am receiving the SNS messages through out the life cycle of the application.
<hr>

> Can you be more descriptive when you say `second print never gets executed and the application stops.`, what does it mean for the application to "stop" ?

if you look at this method 👇
```python
def unsubscribe(self) -> None:
        print("unsubscribe")     # <- first print statement
        res = self.sns_client.unsubscribe(
            SubscriptionArn=self.subscription_arn
        )
        print("suceessfully unsubscribed from topic")     # <- second print statement

```

there are two print statements in this method as indicated by comments `#`

I am calling this method in the FastAPI `shutdown` event

```py
@app.on_event("shutdown")
def app_shutdown():
    unsubscribe()
```

Assuming the app is running perfectly fine, and I press `Ctrl` + `C` (to stop the application that should trigger `shutdown` event [I might be wrong here])

### The expected behavior
 1. the `shutdown` event calls the `unsubscribe` synchronous function
 2. the function execution
     1.  should see `unsubscribe` (first print statement) printed on the console/terminal
     2. the SNS unsubscribe function 
     3. should see `successfully unsubscribed from topic` (second print statement) printed on the console/terminal
     4. event loop stops


### actual behavior
 1. the `shutdown` event calls the `unsubscribe` synchronous function
 2. the function execution
     1.  see `unsubscribe` (first print statement) printed on the console/terminal
     2. event loop stops

<hr>

> `The shutdown handlers will run once all connections have been closed, and any in-process background tasks have completed.` are you holding open a connection or have long running background tasks?

I am not sure what do you mean by open connections but I assume `http` connections, in which case, there is no connection open when I am quitting (`Ctrl + `C`).
I guess the problem here is not that the `shutdown` event is not triggering the function. The problem is that it is not executing the `sync` SNS call before ending the event loop.

<hr>

Hope this clarifies.

**Comment by Kludex** [comment]
Can you provide s minimal, reproducible, example?
